### PR TITLE
Feature/ir links in create view

### PR DESCRIPTION
**Pull request by lamai93**
@vasiliy-arangodb Tests are still missing!

- [ ]  Write proper tests.

Fixes arangodb/team-clifton-planning#12.

### CODE REVIEW

File: arangod/IResearch/IResearchViewCoordinator.cpp [hunk 1]
```diff
@@ -340,11 +340,54 @@ bool IResearchViewCoordinator::emplace(
...
+    auto* engine = arangodb::ClusterInfo::instance();
+
+    if (!engine) {
+      return nullptr;
+    }
+
+    // check link auth as per https://github.com/arangodb/backlog/issues/459
+    if (arangodb::ExecContext::CURRENT) {
+      // check existing links
+      for (auto& entry: view->_collections) {
```
Comment (vasiliy-arangodb) [inline arangod/IResearch/IResearchViewCoordinator.cpp]: why do you expect there to be collections present?

File: arangod/IResearch/IResearchViewCoordinator.cpp [hunk 2]
```diff
@@ -340,11 +340,54 @@ bool IResearchViewCoordinator::emplace(
...
+        auto collection =
+          engine->getCollection(vocbase.name(), std::to_string(entry.first));
+
+        if (collection
+            && !arangodb::ExecContext::CURRENT->canUseCollection(vocbase.name(), collection->name(), arangodb::auth::Level::RO)) {
+          return nullptr;
```
Comment (vasiliy-arangodb) [inline arangod/IResearch/IResearchViewCoordinator.cpp]: should link creation failure cause view creation failure?

File: arangod/IResearch/IResearchViewCoordinator.cpp [hunk 3]
```diff
@@ -340,11 +340,54 @@ bool IResearchViewCoordinator::emplace(
...
+          if (!itr.key().isString()) {
+            continue; // not a resolvable collection (invalid jSON)
+          }
+
+          auto collection =
+            engine->getCollection(vocbase.name(), itr.key().copyString());
+
+          if (collection
+              && !arangodb::ExecContext::CURRENT->canUseCollection(vocbase.name(), collection->name(), arangodb::auth::Level::RO)) {
+            return nullptr;
```
Comment (vasiliy-arangodb) [inline arangod/IResearch/IResearchViewCoordinator.cpp]: should link creation failure cause view creation failure?

File: arangod/IResearch/IResearchViewCoordinator.cpp [hunk 4]
```diff
@@ -397,6 +440,33 @@ bool IResearchViewCoordinator::emplace(
 
       return nullptr;
     }
+
+    // create links - "on a best-effort basis"
+    if (info.hasKey("links")) {
```
Comment (vasiliy-arangodb) [inline arangod/IResearch/IResearchViewCoordinator.cpp]: this should really be in a try-catch since the view definition has already been persisted

File: arangod/IResearch/IResearchViewSingleServer.cpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+#include "VocBase/LogicalView.h"
+#include "Logger/LogMacros.h"
+
+#include "IResearchView.h"
+#include "IResearchViewSingleServer.h"
+
+namespace arangodb {
+namespace iresearch {
+
+/*static*/ std::shared_ptr<LogicalView> IResearchViewSingleServer::make(
```
Comment (vasiliy-arangodb) [inline arangod/IResearch/IResearchViewSingleServer.cpp]: why can the code from IResearchView be reused?

File: arangod/IResearch/IResearchViewSingleServer.h [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+namespace arangodb {
+namespace iresearch {
+
+///////////////////////////////////////////////////////////////////////////////
+/// @class IResearchViewSingleServer
+/// @brief an abstraction over the distributed IResearch index implementing the
+///        LogicalView interface
+///////////////////////////////////////////////////////////////////////////////
+class IResearchViewSingleServer final {
```
Comment (vasiliy-arangodb) [inline arangod/IResearch/IResearchViewSingleServer.h]: Why is this required?

File: arangod/IResearch/IResearchViewCoordinator.cpp (hunk 1 above)
Comment (lamai93) [reply arangod/IResearch/IResearchViewCoordinator.cpp]: This is just copy past error.

File: arangod/IResearch/IResearchViewCoordinator.cpp (hunk 2 above)
Comment (lamai93) [reply arangod/IResearch/IResearchViewCoordinator.cpp]: Should it say that everything is fine?

File: arangod/IResearch/IResearchViewCoordinator.cpp (hunk 3 above)
Comment (lamai93) [reply arangod/IResearch/IResearchViewCoordinator.cpp]: Should it say that everything is fine?

File: arangod/IResearch/IResearchViewCoordinator.cpp (hunk 4 above)
Comment (lamai93) [reply arangod/IResearch/IResearchViewCoordinator.cpp]: Please elaborate on that.

File: arangod/IResearch/IResearchViewSingleServer.cpp (hunk 5 above)
Comment (lamai93) [reply arangod/IResearch/IResearchViewSingleServer.cpp]: Because thats what happens before in the single server case. The only reason I did this, is that the DBServer version reuses the code of the single server.

File: arangod/IResearch/IResearchViewSingleServer.h (hunk 6 above)
Comment (lamai93) [reply arangod/IResearch/IResearchViewSingleServer.h]: Because the DBServer version reuses it. And link creation should only happen in single server mode and on coordinators.

File: arangod/IResearch/IResearchViewCoordinator.cpp (hunk 3 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchViewCoordinator.cpp]: quite possibly

File: arangod/IResearch/IResearchViewCoordinator.cpp (hunk 4 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchViewCoordinator.cpp]: as far as the storage engine is concerned the view exists, so upon recovery the view will be instantiated
if however there is an exception during link creation then the current arangod process will report that the view has not been created, which would be inconsistent with recovery

File: arangod/IResearch/IResearchViewSingleServer.cpp (hunk 5 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchViewSingleServer.cpp]: I am not sure I follow why there needs to be code duplication and the necessity for this class instead of just using IResearchView::make(...)

File: arangod/IResearch/IResearchViewSingleServer.h (hunk 6 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchViewSingleServer.h]: It's valid for a DBServer to use IResearchView::make(...), in the IResearchDBServer::make(...) the links are simply filtered out

**Comment by vasiliy-arangodb** [comment]
It would be nice to see cluster tests

**Comment by lamai93** [comment]
Outdated
